extensions/amp-share-tracking/0.1/amp-share-tracking.js

```diff
@@ -15,41 +15,173 @@
+
+    return Promise.all([
+      this.shareTrackingService_.getIncomingFragment(),
+      this.shareTrackingService_.getOutgoingFragment(this.vendorHref)]).then(
+        results => {
+          this.incomingFragment = results[0];
+          this.outgoingFragment = results[1];
+          dev.fine(TAG, 'get incoming fragment: ' + this.incomingFragment);
+          dev.fine(TAG, 'get outgoing fragment: ' + this.outgoingFragment);
+        });
   }
 }
 
 /**
- * ShareTrackingService processes the incoming and outcoming url fragment
+ * ShareTrackingService processes the incoming and outgoing url fragment
  */
 export class ShareTrackingService {
   constructor(window) {
-    this.win = window;
+    this.win_ = window;
   }
+
+  /**
+   * Get the incoming share-tracking fragment from the url
+   * @return {!Promise<string>}
+   */
+  getIncomingFragment() {
+    // remove the #
+    const hash = this.win_.location.hash.substr(1);
+
+    // when the url contains the share-tracking identifier directly
+    if (hash.indexOf('.') == 0) {
+      const endIndex = hash.indexOf('&');
+      if (endIndex != -1) {
+        return Promise.resolve(hash.substr(1, (endIndex - 1)));
+      } else {
+        return Promise.resolve(hash.substr(1));
+      }
+    }
+
+    // get the share-tracking identifier from the viewer
+    return viewerFor(this.win_).getShareTrackingIncomingFragment();
+  }
+
+  /**
+   * Get an outgoing share-tracking fragment
+   * @return {!Promise<string>}
+   */
+  getOutgoingFragment(vendorUrl) {
+    if (vendorUrl) {
+      return this.getOutgoingFragmentFromVendor(vendorUrl).then(
+        response => {
+          if (!response.outgoingFragment) {
+            dev.warn(TAG, 'The response from [' + vendorUrl +
+                '] does not have an outgoingFragment value. ' +
+                'Generating outgoing fragment using random generator.');
+            return this.getOutgoingFragmentRandomly();
+          } else {
+            return Promise.resolve(response.outgoingFragment);
+          }
+        }
+      );
+    } else {
+      return this.getOutgoingFragmentRandomly();
+    }
+  }
+
+  /**
+   * Get an outgoing share-tracking fragment from vendor
+   * by issueing a post request to the url the vendor provided
+   * @return {!Promise<!ShareTrackingPostResponseDef>}
+   */
+  getOutgoingFragmentFromVendor(vendorUrl) {
+    const postReq = {
+      method: 'POST',
+      credentials: 'include',
+      body: {},
+    };
+    return xhrFor(this.win_).fetchJson(vendorUrl, postReq);
+  }
+
+  /**
+   * Get a random outgoing share-tracking fragment
+   * @return {!Promise<string>}
+   */
+  getOutgoingFragmentRandomly() {
+    const randomFragment = this.getRandomFragment();
+    return Promise.resolve(randomFragment);
+  }
+
+  /**
+   * Get a random fragment
+   * @return {!string}
+   */
+  getRandomFragment() {
+    // TODO(yuxichen): generate random fragment
+    return 'rAmDoM';
```

Comment:
Do you want the random string to be user sticky? If so, you would like to follow what amp-experiment does. A user notification might be needed as well.